anesthetic/utils.py

```diff
@@ -182,3 +183,131 @@ def iso_probability_contours(pdf, contours=[0.68, 0.95]):
                 c[j] = c[j] - 1e-5
 
     return c
+
+
+def iso_probability_contours_from_samples(pdf, contours=[0.68, 0.95],
+                                          weights=None):
+    """Compute the iso-probability contour values."""
+    if weights is None:
+        weights = numpy.ones_like(pdf)
+    contours = [1-p for p in reversed(contours)]
+    i = numpy.argsort(pdf)
+    m = numpy.cumsum(weights[i])
+    m /= m[-1]
+    interp = interp1d([0]+list(m), [0]+list(pdf[i]))
+    c = list(interp(contours))+[max(pdf)]
+
+    # Correct level sets
+    for i in range(1, len(c)):
+        if c[i-1] == c[i]:
+            for j in range(i):
+                c[j] = c[j] - 1e-5
+
+    return c
+
+
+def triangular_sample_compression_2d(x, y, w=None, n=1000):
+    """Histogram a 2D set of weighted samples via triangulation.
+
+    This defines bins via a triangulation of the subsamples, sums weights
+    within triangles, and computes weighted centroids of triangles.
+
+    Parameters
+    ----------
+    x, y: array-like
+        x and y coordinates of samples for compressing
+
+    w: array-like, optional
+        weights of samples
+
+    n: int, optional
+        number of samples returned. Default 1000
+
+    Returns
+    -------
+    x, y, w, array-like
+        Compressed samples and weights
+
+    """
+    if w is None:
+        w = numpy.ones_like(x)
+
+    # Select samples for triangulation
+    if sum(w != 0) < n:
+        i = numpy.arange(len(w))
+    else:
+        i = numpy.random.choice(len(w), size=n, replace=False, p=w/w.sum())
+
+    # Generate triangulation
+    cov = numpy.cov(x, y, aweights=w)
+    L = numpy.linalg.cholesky(cov)
+    Linv = numpy.linalg.inv(L)
+    x_, y_ = Linv.dot([x[i], y[i]])
+    tri = Triangulation(x_, y_)
+
+    # Mask out triangles with unreasonably large perimeters
+    vec = numpy.array([x_[tri.triangles], y_[tri.triangles]]).transpose()
+    s = (numpy.linalg.norm(vec[1, :, :] - vec[0, :, :], axis=1) +
+         numpy.linalg.norm(vec[2, :, :] - vec[1, :, :], axis=1) +
+         numpy.linalg.norm(vec[0, :, :] - vec[2, :, :], axis=1))
+    A = s**2
+    tri.set_mask(A > 25*A.mean())
```

Comment:
Where is the number 25 here coming from? How do you quantify "unreasonably" in "unreasonably large perimeters"?

Reply:
This was very much done by trial and error. 

It combats the convex case (rings or bananas), since then the triangulation has ~O(a few) large perimeter triangles that tile outside the distribution. In the convex case, this line has no effect.

Reply:
This seems a bit arbitrary, could using a z-score or the IQR improve this? 

E.g. crop 1.5 IQR above the 3rd quartile: 
```python
tri.set_mask(s > s.quantile(0.75) + 1.5 * stats.iqr(s))
```

Reply:
Here is a typical histogram of s:
![Figure_1](https://user-images.githubusercontent.com/3708486/62890333-361b7500-bd3b-11e9-92b3-2be491ed354a.png)
We basically want to trim off those points way out in the tail. The quantile iqr approach you suggest would end up trimming too much of the main distribution

Reply:
OK. It seems that taking the log of s, zscoring and removing those points greater than 3 standard deviations seems to do the trick. 3 standard deviations makes sense for 1000 samples, since that's 99.9% of the distribution. Easily generalises to nsamples with scipy.stats.norm.ppf. Changes pushed
![Figure_1](https://user-images.githubusercontent.com/3708486/62891862-bf807680-bd3e-11e9-8d1f-4cc5a99c5f63.png)

Reply:
This looks good. If this also works as desired on banana cases then this is great.

Reply:
Without and with the mask:
![Figure_1](https://user-images.githubusercontent.com/3708486/62899339-7553c100-bd4f-11e9-906d-a3c5fcf4fe2e.png)

Reply:
Oh dear, well with the mask is clearly better... but it doesn't look optimal either...?

Reply:
It's actually quite a difficult one for it to plot. Increasing the number of samples with ncompress improves matters (asking it to plot this one accurately with only 1000 samples is a bit of a tall order) 
![Figure_1](https://user-images.githubusercontent.com/3708486/62916871-106d8a80-bd92-11e9-9e55-61d43b08e2dd.png)

Reply:
Yeah, that looks even better. Is there a way to find out or estimate the minimum amount of `nsamples` needed for neat plots? The `channel_capacity` returns too large numbers which take too long to compute, right?

Reply:
I think this likely requires manual tuning. The main point is that the default should be fast, and good enough. One can always dial up ncompress to maximum once you're happy with the general layout.